code-submissions-clustering-core/src/main/kotlin/org/jetbrains/research/code/submissions/clustering/util/LoadUtil.kt

```diff
@@ -65,16 +80,41 @@ fun SubmissionsGraph.writeToCsv(outputPath: String) {
 }
 
 fun SubmissionsGraph.writeToPng(outputPath: String) {
-    val clustersFilePath = "$outputPath/clusters.png"
+    val visualizationFolder = "$outputPath/visualization"
+    createFolder(visualizationFolder)
+    val clustersFilePath = "$visualizationFolder/clusters.png"
     val clustersFile = File(clustersFilePath)
-    val structureFilePath = "$outputPath/structure.png"
+    val structureFilePath = "$visualizationFolder/structure.png"
     val structureFile = File(structureFilePath)
     visualizeDot(clustersFile, structureFile)
 }
 
-fun SubmissionsGraph.writeClusters(outputPath: String) {
-    val path = "$outputPath/clusters.txt"
+fun SubmissionsGraph.writeClustersToTxt(outputPath: String) {
+    val txtFolder = "$outputPath/txt"
+    createFolder(txtFolder)
+    val path = "$txtFolder/clusters.txt"
     val file = File(path)
     file.createNewFile()
     file.writeText(getClusteredGraph().toString())
 }
+
+private fun SubmissionsGraph.toClusteringDataFrame(): DataFrame<*> {
+    val submissions = mutableListOf<Int>()
+    val clusters = mutableListOf<Int>()
+    val positions = mutableListOf<Int>()
+    getClusteredGraph().graph.vertexSet().forEach { cluster ->
+        var currentPosition = 0
+        cluster.entities.forEach { submissionsNode ->
+            submissionsNode.idList.forEach { submissionId ->
```

Comment:
Should we order solutions by their quality?

Reply:
We should but cannot do that at the moment as we don't store any information about the quality of code submissions yet. Anyway, fixed assigning positions within cluster here to consider all submissions inside the cluster at a time, not only inside one node.